File: sdk/python/gghyper_sdk/pool.py
"""GGCHAIN Mining Pool API client."""
from __future__ import annotations
from dataclasses import dataclass
import requests


@dataclass
class PoolStats:
    hashrate: float
    miners_total: int
    block_height: str
    raw: dict


@dataclass
class MinerStats:
    hashrate: float
    workers_online: int
    workers: dict
    balance: float
    paid: float
    raw: dict


class Pool:
    def __init__(self, base_url: str = "https://pool.gghyper.net"):
        self.base_url = base_url.rstrip("/")

    def _get(self, path: str):
        r = requests.get(f"{self.base_url}{path}", timeout=10)
        r.raise_for_status()
        return r.json()

    def stats(self) -> PoolStats:
        d = self._get("/api/stats")
        node = (d.get("nodes") or [{}])[0]
        return PoolStats(
            hashrate=float(d.get("hashrate") or 0),
            miners_total=int(d.get("minersTotal") or 0),
            block_height=str(node.get("height", "")),
            raw=d,
        )

    def miner(self, address: str) -> MinerStats:
        d = self._get(f"/api/accounts/{address}")
        stats = d.get("stats") or {}
        return MinerStats(
            hashrate=float(d.get("hashrate") or 0),
            workers_online=int(d.get("workersOnline") or 0),
            workers=d.get("workers") or {},
            balance=float(stats.get("balance") or 0) / 1e9,
            paid=float(stats.get("paid") or 0) / 1e9,
            raw=d,
        )

    def blocks(self) -> dict: return self._get("/api/blocks")
    def payments(self) -> dict: return self._get("/api/payments")

    @staticmethod
    def format_hash(h: float) -> str:
        if not h or h <= 0: return "0 H/s"
        units = ["H/s", "KH/s", "MH/s", "GH/s", "TH/s", "PH/s"]
        i = 0
        while h >= 1000 and i < len(units) - 1: h /= 1000; i += 1
        return f"{h:.2f} {units[i]}"

    @staticmethod
    def stratum_url() -> str: return "stratum+tcp://pool.gghyper.net:3333"

File: sdk/js/src/pool.ts
export interface PoolStats {
  hashrate: number;
  minersTotal: number;
  candidatesTotal: number;
  immatureTotal: number;
  maturedTotal: number;
  nodes: Array<{ name: string; height: string; difficulty: string }>;
  now: number;
}

export interface MinerStats {
  hashrate: number;
  hashrate24h?: number;
  currentHashrate?: number;
  workersOnline: number;
  workers: Record<string, { hr: number; lastBeat: number; validShares: number; invalidShares: number }>;
  stats: { balance: number; immature: number; paid: number };
  payments?: Array<{ amount: string; tx: string; timestamp: number }>;
  rewards?: Array<{ blockheight: number; reward: string; timestamp: number }>;
}

/** Client for the GGCHAIN Mining Pool public API. */
export class Pool {
  constructor(public readonly baseUrl: string = "https://pool.gghyper.net") {}

  private async _fetch<T>(path: string): Promise<T> {
    const r = await fetch(`${this.baseUrl}${path}`);
    if (!r.ok) throw new Error(`Pool API ${path}: ${r.status} ${r.statusText}`);
    return await r.json() as T;
  }

  /** Pool-wide live stats. */
  stats(): Promise<PoolStats> { return this._fetch<PoolStats>("/api/stats"); }

  /** Look up a single miner by wallet address. */
  miner(address: string): Promise<MinerStats> { return this._fetch<MinerStats>(`/api/accounts/${address}`); }

  /** Recent blocks found by the pool. */
  blocks(): Promise<any> { return this._fetch("/api/blocks"); }

  /** Recent miner payouts. */
  payments(): Promise<any> { return this._fetch("/api/payments"); }

  /** Helper: format raw H/s into a human string. */
  static formatHash(h: number): string {
    if (!h || h <= 0) return "0 H/s";
    const u = ["H/s", "KH/s", "MH/s", "GH/s", "TH/s", "PH/s"];
    let i = 0; let v = h;
    while (v >= 1000 && i < u.length - 1) { v /= 1000; i++; }
    return `${v.toFixed(2)} ${u[i]}`;
  }

  /** Stratum endpoint string for miners. */
  stratumUrl(): string { return "stratum+tcp://pool.gghyper.net:3333"; }
}
